### PR TITLE
tables: adding chocolatey packages virtual table

### PR DESCRIPTION
**Pull request by muffins**
This brings a new virtual table to Windows, `chocolatey_packages`!  Similar to the python packages table, this lists out all packages which have been installed via chocolatey on a system. It relies on the chocolatey environment variable `ChocolateyInstall` to provide the table with the installation path, after which it pareses the packages `.nuspec`s for information.
```
osquery> .mode line
osquery> select * from chocolatey_packages limit 10;
   name = 7zip.commandline
version = 16.02.0.20170209
summary = 7-Zip is a file archiver with a high compression ratio.
 author = Igor Pavlov
license = http://www.7-zip.org/license.txt
   path = C:\ProgramData\chocolatey\lib\7zip.commandline\7zip.commandline.nuspec

   name = 7zip.portable
version = 16.04
summary = 7-Zip is a file archiver with a high compression ratio.
 author = Igor Pavlov
license = http://www.7-zip.org/license.txt
   path = C:\ProgramData\chocolatey\lib\7zip.portable\7zip.portable.nuspec

   name = aws-sdk-cpp
version = 1.0.107-r1
summary = aws-sdk-cpp
 author = Amazon
license = https://github.com/aws/aws-sdk-cpp/blob/master/LICENSE
   path = C:\ProgramData\chocolatey\lib\aws-sdk-cpp\aws-sdk-cpp.nuspec

   name = boost-msvc14
version = 1.63.0-r2
summary = boost-msvc14
 author = boost-msvc14
license = http://www.boost.org/users/license.html
   path = C:\ProgramData\chocolatey\lib\boost-msvc14\boost-msvc14.nuspec

   name = bzip2
version = 1.0.6
summary = bzip2 libraries and binaries
 author = bzip2
license = http://www.bzip.org/index.html
   path = C:\ProgramData\chocolatey\lib\bzip2\bzip2.nuspec

   name = chocolatey-core.extension
version = 1.3.1
summary = Helper functions extending core choco functionality
 author = chocolatey
license = https://github.com/chocolatey/chocolatey-coreteampackages/blob/master/LICENSE.md
   path = C:\ProgramData\chocolatey\lib\chocolatey-core.extension\chocolatey-core.extension.nuspec

   name = chocolatey-windowsupdate.extension
version = 1.0.2
summary = Helper functions useful for developing packages for Windows updates (KBs).
 author = Jakub Bere┼╝a┼äski
license = https://cdn.rawgit.com/jberezanski/ChocolateyPackages/8086c84fed16d4150a50ba1e97fd6b75e3c4f511/LICENSE
   path = C:\ProgramData\chocolatey\lib\chocolatey-windowsupdate.extension\chocolatey-windowsupdate.extension.nuspec

   name = cmake.portable
version = 3.6.1
summary = Cross-platform, open-source build system including CMake, CTest, CPack, and CMake-GUI
 author = Andy Cedilnik, Bill Hoffman, Brad King, Ken Martin, Alexander Neundorf
license = http://www.cmake.org/licensing/
   path = C:\ProgramData\chocolatey\lib\cmake.portable\cmake.portable.nuspec

   name = cpp-netlib
version = 0.12.0-r4
summary = cpp-netlib
 author = cpp-netlib
license = https://github.com/cpp-netlib/cpp-netlib/blob/master/LICENSE_1_0.txt
   path = C:\ProgramData\chocolatey\lib\cpp-netlib\cpp-netlib.nuspec

   name = cppcheck
version = 1.79.0
summary = A tool for static C/C++ code analysis
 author = The CPPCheck Team
license = http://www.gnu.org/copyleft/gpl.html
   path = C:\ProgramData\chocolatey\lib\cppcheck\cppcheck.nuspec
```

### CODE REVIEW

File: osquery/tables/system/windows/chocolatey_packages.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+namespace osquery {
+namespace tables {
+
+Status genPackage(fs::path nuspec, Row& r) {
+  std::string content;
+  if (!readFile(nuspec, content).ok()) {
+    return Status(1, "Failed to read nuspec:" + nuspec.string());
+  }
+
+  std::stringstream ss;
```
Comment (theopolis) [inline osquery/tables/system/windows/chocolatey_packages.cpp]: Place this into it's own scope:
```cpp
pt::ptree propTree;
{
  std::stringstream ss;
  ss << content;
  read_xml(ss, propTree);
}
```
You could move the `std::string` and `readFile` into this limited context too. Outside of the scope you'll allow the string and string stream to be deallocated. Such that the stringstream member buffer, string buffer, and property tree buffer do not all exists together.

File: osquery/tables/system/windows/chocolatey_packages.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  ss << content;
+  pt::ptree propTree;
+  read_xml(ss, propTree);
```
Comment (theopolis) [inline osquery/tables/system/windows/chocolatey_packages.cpp]: This may throw, do you want to handle it here, or allow the registry to catch it?

Comment (theopolis) [reply osquery/tables/system/windows/chocolatey_packages.cpp]: If you handle it, and return a 1 status you will allow the table to emit partial results.

File: osquery/tables/system/windows/chocolatey_packages.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+  auto chocoEnvInstall = getEnvVar("ChocolateyInstall");
+
+  fs::path chocoInstallPath;
+  if (chocoEnvInstall.is_initialized()) {
+    chocoInstallPath = fs::path(*chocoEnvInstall);
+  }
+
+  if (chocoInstallPath.empty()) {
+    LOG(WARNING) << "Did not find chocolatey path environment variable.";
```
Comment (theopolis) [inline osquery/tables/system/windows/chocolatey_packages.cpp]: nit, remove the period.

File: osquery/tables/system/windows/chocolatey_packages.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+#include "osquery/core/conversions.h"
+#include "osquery/core/process.h"
+
+namespace fs = boost::filesystem;
+namespace pt = boost::property_tree;
+
+namespace osquery {
+namespace tables {
+
+Status genPackage(fs::path nuspec, Row& r) {
```
Comment (theopolis) [inline osquery/tables/system/windows/chocolatey_packages.cpp]: `const fs::path& nuspec`

File: osquery/tables/system/windows/chocolatey_packages.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+#include <boost/filesystem.hpp>
+#include <boost/property_tree/xml_parser.hpp>
+
+#include <osquery/core.h>
+#include <osquery/filesystem.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include "osquery/core/conversions.h"
+#include "osquery/core/process.h"
```
Comment (theopolis) [inline osquery/tables/system/windows/chocolatey_packages.cpp]: Why do you need these two (process, conversions) includes?

Comment (muffins) [reply osquery/tables/system/windows/chocolatey_packages.cpp]: Ah, `process.h` was for `getEnvVar`. I was using `osquery::split` earlier but found it wasn't needed, and forgot to remove the header. Thanks!

File: osquery/tables/system/windows/chocolatey_packages.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+#include "osquery/core/process.h"
+
+namespace fs = boost::filesystem;
+namespace pt = boost::property_tree;
+
+namespace osquery {
+namespace tables {
+
+Status genPackage(const fs::path nuspec, Row& r) {
```
Comment (theopolis) [inline osquery/tables/system/windows/chocolatey_packages.cpp]: Should `nuspec` be passed by reference?

**Comment by osqueryer** [comment]
:no_good: The commit 01eb4071ab80f4b22113291b9b28a6f3ef0e58b9 (Job results: [2341](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2341/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 01eb4071ab80f4b22113291b9b28a6f3ef0e58b9 (Job results: [5653](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5653/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 01eb4071ab80f4b22113291b9b28a6f3ef0e58b9 (Job results: [2539](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2539/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 01eb4071ab80f4b22113291b9b28a6f3ef0e58b9 (Job results: [2541](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2541/)) failed one or more tests (**Windows**).

**Pull request merged**